### PR TITLE
6099 add spacy pyfunc example

### PR DESCRIPTION
**Pull request by dipanjank**
## Related Issues/PRs

 #6099 

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)
This PR adds an example of how to 

* train a custom Spacy ``TextCategorizer`` model 
* capture the historical training loss
* log the artifact with mlflow and retrieve it later

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [X] `area/docs`: MLflow documentation pages
- [X] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: docs/source/models.rst [hunk 1]
```diff
@@ -1122,6 +1122,194 @@ in native spaCy format.
...
+~~~~~~~~~~~~~~~~~~~~~~
+
+The example below shows how to train a ``Spacy`` ``TextCategorizer`` model, log the model artifact and metrics to the
+mlflow tracking server and then load the saved model to make predictions. For this example, we will be using the
+``Polarity 2.0`` dataset available in the ``nltk`` package. This dataset consists of 10000 positive and 10000 negative
+short movie reviews.
+
+First we convert the texts and sentiment labels ("pos" or "neg") from NLTK native format to ``Spacy``'s ``DocBin`` format:
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (dipanjank) [reply docs/source/models.rst:1135]: Hello! I understand that this example is a bit longer than most others. IMO, this is because Spacy has moved to a largely command line + config driven user interface compared to a programmatic user interface like pytorch and tensorflow. So the only way to have a single code script that does all of the things you mention would be to copy & paste part of Spacy's interface. Please see here: https://spacy.io/usage/training#quickstart

The only way I can think of making the example simpler is to not use a Text Classification problem, e.g. https://github.com/arnaudsj/sample-mlflow-spacy/blob/master/train_predict.py#L28-L39. This will remove the need for the data processing functions. I'm happy to do that, but IMO the text classification problem is something that we see in practice the most, so it's definite more useful for mlflow users. Please LMK which one you prefer.

[the remark this answers is not in the document]
Comment (harupy) [reply docs/source/models.rst:1135]: @dipanjank Understood! Didn't know Spacy has moved to a largely command line + config driven user interface. Given that, the current example looks reasonable.

File: docs/source/models.rst [hunk 2]
```diff
@@ -1122,6 +1122,194 @@ in native spaCy format.
 
 For more information, see :py:mod:`mlflow.spacy`.
 
+``Spacy`` pyfunc usage
+~~~~~~~~~~~~~~~~~~~~~~
+
+The example below shows how to train a ``Spacy`` ``TextCategorizer`` model, log the model artifact and metrics to the
+mlflow tracking server and then load the saved model to make predictions. For this example, we will be using the
+``Polarity 2.0`` dataset available in the ``nltk`` package. This dataset consists of 10000 positive and 10000 negative
```
Comment (harupy) [inline docs/source/models.rst:1130]: Does this example finish quickly? Curious how long it takes to preprocess data and train a model.

File: docs/source/models.rst [hunk 3]
```diff
@@ -1122,6 +1122,194 @@ in native spaCy format.
...
+    train = null
+    dev = null
+
+to the current directory:
+
+.. code-block:: text
+
+    [paths]
+    train = "."
+    dev = "."
```
Comment (harupy) [inline docs/source/models.rst:1213]:
```suggestion
Change the default train and dev paths in the config file to the current directory:

.. code-block:: diff

      [paths]
    - train = null
    - dev = null
    + train = "."
    + dev = "."
```

File: docs/source/models.rst [hunk 4]
```diff
@@ -1122,6 +1122,194 @@ in native spaCy format.
...
+    [training.logger]
+    @loggers = "spacy.ConsoleLogger.v1"
+    progress_bar = false
+
+section of the config file to
+
+.. code-block:: text
+
+    [training.logger]
+    @loggers = "mlflow_logger.v1"
```
Comment (harupy) [inline docs/source/models.rst:1282]: Can we apply the same fix as https://github.com/mlflow/mlflow/pull/7753/files#r1092761717 here?

File: docs/source/models.rst [hunk 5]
```diff
@@ -1122,6 +1122,194 @@ in native spaCy format.
...
+    def convert(data_df: pd.DataFrame, target_file: str):
+        """Convert a DataFrame with 'sentence' and 'sentiment' columns to a
+        spacy DocBin object and save it to 'target_file'.
+        """
+        nlp = spacy.blank("en")
+        sentiment_labels = data_df.sentiment.unique()
+        spacy_doc = DocBin()
+
+        for _, row in data_df.iterrows():
+            sent_tokens = nlp.make_doc(row.sentence)
```
Comment (harupy) [inline docs/source/models.rst:1167]:
```suggestion
            sent_tokens = nlp.make_doc(row["sentence"])
```

File: docs/source/models.rst [hunk 6]
```diff
@@ -1122,6 +1122,194 @@ in native spaCy format.
...
+            # To train a Spacy TextCategorizer model, the label must be attached to the "cats" dictionary of the "Doc"
+            # object, e.g. {"pos": 1.0, "neg": 0.0} for a "pos" label.
+            for label in sentiment_labels:
+                sent_tokens.cats[label] = 1.0 if label == row.sentiment else 0.0
```
Comment (harupy) [inline docs/source/models.rst:1171]:
```suggestion
                sent_tokens.cats[label] = 1.0 if label == row["sentiment"] else 0.0
```

File: docs/source/models.rst [hunk 7]
```diff
@@ -1122,6 +1122,194 @@ in native spaCy format.
...
+    # We need to append the spacy model directory name to the artifact uri
+    spacy_model = mlflow.pyfunc.load_model(
+        "/".join([last_run.info.artifact_uri, "mlflow_textcat_example"])
+    )
+    predictions_in = dev_df.loc[:, ["sentence"]]
+    predictions_out = spacy_model.predict(predictions_in).squeeze().tolist()
+    predicted_labels = [
+        "pos" if row["pos"] > row["neg"] else "neg" for row in predictions_out
+    ]
+    pred_total = dev_df.assign(predicted_sentiment=predicted_labels)
```
Comment (harupy) [inline docs/source/models.rst:1311]:
```suggestion
    print(dev_df.assign(predicted_sentiment=predicted_labels))
```

File: docs/source/models.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (dipanjank) [reply docs/source/models.rst:1135]: Cool, then I will fix the existing review comments in this PR.

File: docs/source/models.rst (hunk 2 above)
Comment (dipanjank) [reply docs/source/models.rst:1130]: It takes about 10 minutes to train this particular with the default number of training steps in the spacy config file (2000).

File: docs/source/models.rst (hunk 4 above)
Comment (dipanjank) [reply docs/source/models.rst]: Fixed.

**Comment by dipanjank** [comment]
@harupy  can you please review again? I applied all your review comments?

File: docs/source/models.rst (hunk 2 above)
Comment (harupy) [reply docs/source/models.rst:1130]: 10 minutes is long. Can we reduce the training data size or the number of training steps?

Comment (dipanjank) [reply docs/source/models.rst:1130]: Sure, added the instruction to reduce the number of training steps to 100 (default is 20000, not 2000) and the eval frequency to 10 (default is 100). Now it completes in 1-2 minutes.

Comment (dipanjank) [reply docs/source/models.rst:1130]: https://github.com/mlflow/mlflow/pull/7753/files#diff-27092883ef9a4a81ab9b5ea59442ab48dd314babd61428270948fcfae0a80f28R1273-R1277

File: docs/source/models.rst [hunk 8]
```diff
@@ -1122,6 +1122,189 @@ in native spaCy format.
...
+        lambda x: x.sample(frac=0.7, random_state=100)
+    )
+    dev_df = review_data.loc[review_data.index.difference(train_df.index), :]
+
+    # Save the train and dev data files to the current directory as "corpora.train" and "corpora.dev", respectively
+    convert(train_df, "corpora.train")
+    convert(dev_df, "corpora.dev")
+
+
+To set up the training job, we first need to generate a configuration file as described in the `Spacy Documentation <https://spacy.io/usage/training#config>` _
```
Comment (harupy) [inline docs/source/models.rst:1192]:
```suggestion
To set up the training job, we first need to generate a configuration file as described in the `Spacy Documentation <https://spacy.io/usage/training#config>`_
```

File: docs/source/models.rst [hunk 9]
```diff
@@ -1122,6 +1122,189 @@ in native spaCy format.
...
+
+    from mlflow import MlflowClient
+
+    # look up the last run info from mlflow
+    client = MlflowClient()
+    last_run = client.search_runs(experiment_ids=["0"], max_results=1)[0]
+
+    # We need to append the spacy model directory name to the artifact uri
+    spacy_model = mlflow.pyfunc.load_model(
+        "/".join([last_run.info.artifact_uri, "mlflow_textcat_example"])
```
Comment (harupy) [inline docs/source/models.rst:1299]:
```suggestion
        f"{last_run.info.artifact_uri}/mlflow_textcat_example"
```

I think f-string is more straightforward in this case.

File: docs/source/models.rst (hunk 2 above)
Comment (harupy) [reply docs/source/models.rst:1130]: > Now it completes in 1-2 minutes.

Great!

**Comment by dipanjank** [comment]
@harupy updated with your last two suggestions.

**Comment by harupy** [comment]
@dipanjank Can you fix the DCO check failure?

**Review by harupy — APPROVED** [review]
LGTM!
